i2c: Update terminology to match i2c spec

teburd opened the pull request:
Updates the terms from master to controller, and slave to target
along with all APIs and usage of those APIs.

The new terminology matches well with our coding guidelines of inclusive
language.

This does not rename internal terminology used in each driver but all
public facing APIs, types, and macros use the new terms.

Fixes #27033

Commit:
docs: i2c: Documentation terminology updated

The i2c terminology has been updated as such master is now controller,
and slave is now target. Updates all doc comments and doc pages to use
updated terminology. Does not change types or API definitions.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
i2c: Update API terminology

Updates the API and types to match updated I2C terminology. Replaces master
with controller and slave with target.

Updates all drivers to match the changed macros, types, and API signatures.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

teburd commented:
> Do we need to keep the old Kconfigs and API wrappers, and marking them as deprecated (as described in our API lifecycle doc)?

Probably!

carlescufi commented:
@teburd can you rebase this please?

carlescufi commented on include/zephyr/drivers/i2c.h:
```diff
@@ -58,8 +58,8 @@ extern "C" {
 
-/** Controller to act as Controller. */
-#define I2C_MODE_MASTER			BIT(4)
+/** Peripheral to act as Controller. */
+#define I2C_MODE_CONTROLLER		BIT(4)
```
Indeed, `I2C_MODE_MASTER` should be deprecated @teburd

carlescufi reviewed (changes requested):
Missing deprecation

henrikbrixandersen commented on drivers/i2c/i2c_dw.c:
```diff
@@ -275,8 +275,8 @@ static void i2c_dw_isr(const struct device *port)
 
 	} else {
-#ifdef CONFIG_I2C_SLAVE
-		const struct i2c_slave_callbacks *slave_cb = dw->slave_cfg->callbacks;
```
General comment: Should we rename the `struct i2c_target_callback` instances as well all over?

teburd replied:
Yes absolutely, but this can be done over time by myself or driver authors, the goal in this PR is very explicit in correcting the public facing terminology. The drivers will require additionally combing over and cleaning up. A task can be created to cover that and it can be delegated to various driver authors hopefully.

henrikbrixandersen commented on drivers/i2c/i2c_lpc11u6x.c:
```diff
@@ -349,8 +349,8 @@ static int lpc11u6x_i2c_init(const struct device *dev)
 	.transfer = lpc11u6x_i2c_transfer,
-	.slave_register = lpc11u6x_i2c_slave_register,
-	.slave_unregister = lpc11u6x_i2c_slave_unregister,
+	.target_register = lpc11u6x_i2c_slave_register,
+	.target_unregister = lpc11u6x_i2c_slave_unregister,
 };
```
Should we rename the driver functions here as well?

gmarull replied:
good point, I think for consistency we should.

teburd replied:
Yes, but that could be done gradually over time, the goal here was to correct all the public facing APIs as noted in the PR

henrikbrixandersen commented on doc/hardware/peripherals/i2c.rst:
```diff
@@ -6,50 +6,41 @@ I2C
 Overview
 ********
 
-.. note::
-
-   Zephyr recognizes the need to change the terms "master" and "slave"
-   used in the current `I2C Specification <i2c-specification>`_.  This
-   will be done when the conditions identified in
-   :ref:`coding_guideline_inclusive_language` have been met.  Existing
-   documentation, data structures, functions, and value symbols in code
-   are likely to change at that point.
-
```
I think it would beneficial to our readers to keep some sort of clarification to the terminology used within the Zephyr I2C controller/target APIs. People unfamiliar with the recent change of terminology in the I2C specification can have difficulties understanding it otherwise.

teburd replied:
This makes sense, I can add a new note pointing out that the terminology has been updated and redefine what a controller and target actually mean. I have no interest in directly mapping the old terminology to the new terminology, as that continues the propagation of it.

henrikbrixandersen commented on doc/hardware/peripherals/i2c.rst:
```diff
@@ -8,52 +8,48 @@ Overview
-   :ref:`coding_guideline_inclusive_language` have been met.  Existing
-   documentation, data structures, functions, and value symbols in code
-   are likely to change at that point.
+   The terminology used in the specification has changed. See
+   https://github.com/zephyrproject-rtos/zephyr/issues/27033
```
I'd prefer a reference to the specification instead of a Zephyr issue. 

```suggestion
   The terminology used in the Zephyr I2C APIs follows that of the `NXP I2C-bus specification and user manual (UM10204), Rev. 7.0 <https://www.nxp.com/docs/en/user-guide/UM10204.pdf>`_. 
```

teburd commented:
> Missing deprecation

Fixed

henrikbrixandersen reviewed (dismissed):
@teburd Thank you for all your work on this!

teburd commented:
The missing copyright failure from CI is not a change related to this PR and I'm happy to add a copyright if that makes sense, I'm just not sure what that should look like.

MaureenHelm commented on include/zephyr/drivers/i2c.h:
```diff
@@ -55,11 +55,14 @@ extern "C" {
 						>> I2C_SPEED_SHIFT)
 
-/** Use 10-bit addressing. DEPRECATED - Use I2C_MSG_ADDR_10_BITS instead. */
+/** @deprecated Use I2C_MSG_ADDR_10_BITS instead. */
 #define I2C_ADDR_10_BITS		BIT(0)
```
Add `__DEPRECATED_MACRO`

MaureenHelm commented on doc/develop/api/overview.rst:
```diff
@@ -162,7 +162,7 @@ current :ref:`stability level <api_lifecycle>`.
      - 3.1
 
-   * - :ref:`i2c-slave-api`
+   * - :ref:`i2c-target-api`
```
Update the version modified entry to 3.2

MaureenHelm reviewed (changes requested):
Needs rebase.

A couple of nits, but otherwise looks great.

Commit:
docs: Update i2c API modified version to 3.2

Updates the API overview page describing that the i2c API has been modified
for 3.2.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

cfriedt commented:
@teburd - looks good, just missing one copyright notice.

teburd commented:
@cfriedt that was missing before, I can’t simply claim it
